src/main/java/org/jabref/gui/entryeditor/SourceTab.java

```diff
@@ -0,0 +1,249 @@
+        inputMap.put(Globals.getKeyPrefs().getKey(KeyBinding.ENTRY_EDITOR_PREVIOUS_PANEL_2), "left");
+        actionMap.put("left", switchLeftAction);
+
+        inputMap.put(Globals.getKeyPrefs().getKey(KeyBinding.HELP), "help");
+        actionMap.put("help", helpAction);
+
+        inputMap.put(Globals.getKeyPrefs().getKey(KeyBinding.NEXT_TAB), "nexttab");
+        actionMap.put("nexttab", frame.nextTab);
+        inputMap.put(Globals.getKeyPrefs().getKey(KeyBinding.PREVIOUS_TAB), "prevtab");
+        actionMap.put("prevtab", frame.prevTab);
+
+        Set<AWTKeyStroke> keys = new HashSet<>(
+                textComponent.getFocusTraversalKeys(KeyboardFocusManager.FORWARD_TRAVERSAL_KEYS));
+        keys.clear();
+        keys.add(AWTKeyStroke.getAWTKeyStroke("pressed TAB"));
+        textComponent.setFocusTraversalKeys(KeyboardFocusManager.FORWARD_TRAVERSAL_KEYS, keys);
+        keys = new HashSet<>(textComponent
+                .getFocusTraversalKeys(KeyboardFocusManager.BACKWARD_TRAVERSAL_KEYS));
+        keys.clear();
+        keys.add(KeyStroke.getKeyStroke("shift pressed TAB"));
+        textComponent.setFocusTraversalKeys(KeyboardFocusManager.BACKWARD_TRAVERSAL_KEYS, keys);
+
+        textComponent.addFocusListener(new EntryEditor.FieldListener());
+        */
+    }
+
+    private void storeSource() {
+        if (codeArea.getText().isEmpty()) {
+            return;
+        }
+
+        BibtexParser bibtexParser = new BibtexParser(Globals.prefs.getImportFormatPreferences());
+        try {
+            ParserResult parserResult = bibtexParser.parse(new StringReader(codeArea.getText()));
+            BibDatabase database = parserResult.getDatabase();
+
+            if (database.getEntryCount() > 1) {
+                throw new IllegalStateException("More than one entry found.");
+            }
+
+            if (!database.hasEntries()) {
+                if (parserResult.hasWarnings()) {
+                    // put the warning into as exception text -> it will be displayed to the user
+                    throw new IllegalStateException(parserResult.warnings().get(0));
+                } else {
+                    throw new IllegalStateException("No entries found.");
+                }
+            }
+
+            NamedCompound compound = new NamedCompound(Localization.lang("source edit"));
+            BibEntry newEntry = database.getEntries().get(0);
+            String newKey = newEntry.getCiteKeyOptional().orElse(null);
+            boolean entryChanged = false;
+            boolean emptyWarning = (newKey == null) || newKey.isEmpty();
+
+            if (newKey != null) {
+                entry.setCiteKey(newKey);
+            } else {
+                entry.clearCiteKey();
+            }
+
+            // First, remove fields that the user has removed.
+            for (Map.Entry<String, String> field : entry.getFieldMap().entrySet()) {
```

Comment:
You mean as entrySet().stream().foreach(...) ? I really like the old-style for loop more when it comes to a complicated body and find foreach only preferable when calling one method.